src/logic/graphFrames.ts

```diff
@@ -7,70 +7,98 @@ import {
   GraphIndex,
 } from './commonInterfaces';
 
+
+// TODO: Refactor the function chain
 interface FrameWithEdges {
-  activeEdges: Array<Edge>,
-  inactiveEdges: Array<Edge>,
+  processedEdges: Array<Edge>,
+  unvisitedEdges: Array<Edge>,
 };
 
+interface FrameWithEdgesAndNodes extends FrameWithEdges{
+  visitedNodes: Array<number>,
+  activeNode: number
+}
+
 export default function createFrames(
   dissasembledGraph: DisassembledGraph,
   algorithm: Algorithm,
 ):
   Array<Frame>
 {
-  const edgeFrames = createFramesWithEdges(algorithm);
-  return convertToIndexes(dissasembledGraph, edgeFrames);
-}
-
-export function createFramesWithEdges(from: Algorithm) {
-  let frames: Array<FrameWithEdges> = [];
-  for (let travelsal of from.traversalHistory) {
-    frames.push(createFrameForNode(frames));
-    frames.push(...createFramesForLookups(travelsal, frames));
+  let frames: Array<FrameWithEdgesAndNodes> = [];
+  const visitedNodes : Array<number> = [];
+  for (let travelsal of algorithm.traversalHistory) {
+    const framesWithEdgeData = [
+      createFrameForNode(frames), ...createFramesForLookups(travelsal, frames),
+    ];
+    frames.push(...addNodeFrameData(framesWithEdgeData, travelsal.node, visitedNodes));
+    visitedNodes.push(travelsal.node);
```

Comment:
visited nodes are kept in frame, why there is separate data structure to keep this?

Reply:
Just pass in visited nodes from last frame to `addNodeFrameData`

Reply:
could u show me what u mean? I don't see it

Reply:
You use array called `visitedNodes` here. This is unnecessary. You could get this information from `lastFrame.visitedNodes`. It's just a matter of selecting lastFrame. So line 34 could look something like this:
`frames.push(...addNodeFrameData(framesWithEdgeData, travelsal.node, lastFrame.visitedNodes));`

Reply:
You're right, but implementing this we would output a real sculpture as a result.  

Look at line 35, I need to append the current node after each traversal. Also, initially, the frames are empty. Indeed we add an additional data structure but I think this is the most straightforward way to do it. 

If u insist to change it, please paste here a working snippet and if it's more articulate than my solution I'll use it. (I've tried to go with ur idea but it was not clearer)